src/browser/cached_resource_fetcher.ts

```diff
@@ -13,85 +13,213 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    const fetcher: Fetcher = isURL(url) ? netRequester : fileRequester;
+
+    return fetcher(url, encoding);
+}
+
+function fileRequester(url: string, encoding: string): Promise<FetchResponse> {
+    return new Promise((resolve, reject) => {
+        // remove possible URI (file:/// scheme) prefix
+        const filepath: string = url.replace(Patterns.URI, '');
+        const isImage = isImageFile(url) || isICO(url);
+
+        if (isImage) {
+            encoding = 'binary';
+        }
+
+        readFile(filepath, encoding, (error: Error | string, data: string) => {
+            if (error) {
+                if (/ENOENT/.test((<Error>error).message || <string>error)) {
+                    resolve(new FetchResponse(false));
                 } else {
-                    callback(null, fileUrl);
+                    reject(error);
                 }
-            });
+            } else {
+                const fetchResponse: FetchResponse = new FetchResponse(true, data, isImage);
+
+                if (isImage) {
+                    fetchResponse.getNativeImage = () => fetchResponse.getNativeImageFromPath(filepath);
+                }
+
+                resolve(fetchResponse);
+            }
+        });
+    });
+}
+
+function netRequester(url: string, encoding: string): Promise<FetchResponse> {
+    return new Promise((resolve, reject) => {
+        const isImage = isImageFile(url) || isICO(url);
+        const request = net.request(url);
+
+        if (isImage) {
+            encoding = 'binary';
         }
-        return;
-    }
 
-    const appCacheDir = getAppCacheDir(appUuid);
-    const filePath = getFilePath(appCacheDir, fileUrl);
+        request.on('error', reject); // this is an error making the request
+
+        request.on('response', (response: ClientResponse) => {
+            const chunks: string[] = [];
+            const fetchResponse: FetchResponse = new FetchResponse(response.statusCode === 200, undefined, isImage);
+            if (isImage) {
+                fetchResponse.getNativeImage = () => fetchResponse.getNativeImageFromCreatedPath(url);
+            }
+
+            if (!fetchResponse.success) {
+                resolve(fetchResponse); // not an error, however `success` will be false and `data` will be undefined
+                return;
+            }
+
+            response.on('error', reject); // this is an error receiving the response
+
+            response.setEncoding(encoding);
+            response.on('data', (chunk: string) => {
+                chunks.push(chunk);
+            });
+
+            response.on('end', () => {
+                fetchResponse.data = chunks.join('');
+                resolve(fetchResponse);
+            });
+        });
+
+        request.end();
+    });
+}
 
-    stat(filePath, (err: null | Error) => {
-        if (err) {
-            if (!appQuiting) {
-                stat(appCacheDir, (err: null | Error) => {
+function makeAppCacheDir(dir: string): Promise<string> {
```

Comment:
:neckbeard:  if you pass in the `appCacheDir` this function would be a lot more testable (and general)

Reply:
not sure I understand you:
not sure how that would help... we still need to check if it exists and if not create it....